test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs

```diff
@@ -0,0 +1,647 @@
+    {
+        static List<LogEntry> CompleteLogEntries;
+        static Dictionary<string, LogLevel> LogEntries;
+
+        static OpenIdConnectHandlerTests()
+        {
+            LogEntries =
+                new Dictionary<string, LogLevel>()
+                {
+                    { "OIDCH_0000:", LogLevel.Debug },
+                    { "OIDCH_0001:", LogLevel.Debug },
+                    { "OIDCH_0002:", LogLevel.Information },
+                    { "OIDCH_0003:", LogLevel.Information },
+                    { "OIDCH_0004:", LogLevel.Error },
+                    { "OIDCH_0005:", LogLevel.Error },
+                    { "OIDCH_0006:", LogLevel.Error },
+                    { "OIDCH_0007:", LogLevel.Error },
+                    { "OIDCH_0008:", LogLevel.Debug },
+                    { "OIDCH_0009:", LogLevel.Debug },
+                    { "OIDCH_0010:", LogLevel.Error },
+                    { "OIDCH_0011:", LogLevel.Error },
+                    { "OIDCH_0012:", LogLevel.Debug },
+                    { "OIDCH_0013:", LogLevel.Debug },
+                    { "OIDCH_0014:", LogLevel.Debug },
+                    { "OIDCH_0015:", LogLevel.Debug },
+                    { "OIDCH_0016:", LogLevel.Debug },
+                    { "OIDCH_0017:", LogLevel.Error },
+                    { "OIDCH_0018:", LogLevel.Debug },
+                    { "OIDCH_0019:", LogLevel.Debug },
+                    { "OIDCH_0020:", LogLevel.Debug },
+                    { "OIDCH_0026:", LogLevel.Error },
+                };
+
+            BuildLogEntryList();
+        }
+
+        private static void BuildLogEntryList()
+        {
+            CompleteLogEntries = new List<LogEntry>();
+            foreach(var entry in LogEntries)
+            {
+                CompleteLogEntries.Add(new LogEntry { State = entry.Key, Level = entry.Value });
+            }
+        }
+
+        [Fact]
+        public void TestLoggingMask()
+        {
+            var logger = new Logger(LogLevel.Debug);
+            logger.IsEnabled(LogLevel.Critical).ShouldBe<bool>(true);
+            logger.IsEnabled(LogLevel.Debug).ShouldBe<bool>(true);
+            logger.IsEnabled(LogLevel.Error).ShouldBe<bool>(true);
+            logger.IsEnabled(LogLevel.Information).ShouldBe<bool>(true);
+            logger.IsEnabled(LogLevel.Verbose).ShouldBe<bool>(true);
+            logger.IsEnabled(LogLevel.Warning).ShouldBe<bool>(true);
+
+            logger = new Logger(LogLevel.Critical);
+            logger.IsEnabled(LogLevel.Critical).ShouldBe<bool>(true);
+            logger.IsEnabled(LogLevel.Debug).ShouldBe<bool>(false);
+            logger.IsEnabled(LogLevel.Error).ShouldBe<bool>(false);
+            logger.IsEnabled(LogLevel.Information).ShouldBe<bool>(false);
+            logger.IsEnabled(LogLevel.Verbose).ShouldBe<bool>(false);
+            logger.IsEnabled(LogLevel.Warning).ShouldBe<bool>(false);
+        }
+
+        [Fact]
+        public async Task AuthenticateCore()
```

Comment:
1. Should use a `[Theory]` for this to test multiple variations
2. Test names should say what they're doing and what they're expecting. E.g. `AuthenticateCore_WithXyz_DoesSuchAndSuch`.